subprojects/enterprise-operations/src/main/java/org/gradle/execution/work/WorkExecutionBuildOperationType.java

```diff
@@ -0,0 +1,129 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.execution.work;
+
+import org.gradle.internal.operations.BuildOperationType;
+
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Captures the execution of a unit of work. This includes all operations related to Gradle's bookkeeping and optimizations.
+ * Among other things it includes snapshotting inputs and outputs, up-to-date checks, build cache operations.
+ * It also wraps the {@link WorkActionExecutionBuildOperationType execution of the action}.
+ *
+ * The operations can be nested, i.e. when a task executes a worker API action.
+ *
+ * @see WorkActionExecutionBuildOperationType
+ */
+public interface WorkExecutionBuildOperationType extends BuildOperationType<WorkExecutionBuildOperationType.Details, WorkExecutionBuildOperationType.Result> {
+
+    interface Details {
+
+        /**
+         * The type of work represented by the build operation.
+         */
+        Class<?> getWorkType();
+
+        /**
+         * Path to the owning build.
+         */
+        String getBuildPath();
+
+        /**
+         * A textual identifier for the unit of work when available.
+         *
+         * Tasks will expose their task paths here. Artifact transforms and workers don't expose anything here.
+         */
+        Optional<String> getIdentity();
+
+        /**
+         * A unique number identifying the unit of work being executed in the context of the current build.
+         */
+        long getUniqueId();
+
+        /**
+         * The type implementing the unit of work.
+         */
+        Class<?> getImplementation();
+    }
+
+    interface Result {
+        /**
+         * Whether the unit of work had any actions. Lifecycle tasks don't have actions.
+         *
+         * @see org.gradle.api.internal.tasks.TaskStateInternal#isActionable()
+         */
+        boolean isActionable();
+
+        /**
+         * Returns if this unit of work was executed incrementally.
+         *
+         * @see org.gradle.work.InputChanges#isIncremental()
+         */
+        boolean isIncremental();
+
+        /**
+         * The message describing why the unit of work was skipped entirely, {@link Optional#empty()} if the task wasn't skipped.
+         */
+        Optional<String> getSkipMessage();
+
+        /**
+         * Opaque messages describing why the work was not up-to-date.
+         * In the order emitted by Gradle.
+         * {@link Optional#empty()} if execution did not get so far as to test “up-to-date-ness”.
+         * Empty list if tested, but task was considered up-to-date.
+         */
```

Comment:
` Optional#empty()} if execution did not get so far as to test “up-to-date-ness”.` 

Would that information deserve a separate field, something like `workStatus` or `workPhase`.  It feels dirty to incorporate information if execution got as far the test “up-to-date-ness” in to the `getUpToDateMessages()`.

Reply:
Note that I tried to preserve the structure of `ExecuteTaskBuildOperationType` as much as possible. We can go into refactoring this, but that adds cost. Also, if we introduce something like `Optional<CacheReuseInfo>` to hold all the data about reusing a previous result (either by being up-to-date or by loading from cache), adding more structure might mean worse adaptability in future.

For now I opted to keep the changes minimal.